service-tests/xml.js

```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const Joi = require('joi');
+const ServiceTester = require('./runner/service-tester');
+const {
+  isSemver,
+} = require('./helpers/validators');
+
+const colorscheme = require('../lib/colorscheme.json');
+const mapValues = require('lodash.mapvalues');
+
+const colorsB = mapValues(colorscheme, 'colorB');
+
+const t = new ServiceTester({ id: 'badge/dynamic/xml', title: 'User Defined XML Source Data' });
+module.exports = t;
+
+t.create('Connection error')
+  .get('.json?uri=https://services.addons.mozilla.org/en-US/firefox/api/1.5/addon/707078&query=/addon/name&label=Package Name&style=_shields_test')
+  .networkOff()
+  .expectJSON({ name: 'Package Name', value: 'inaccessible', colorB: colorsB.red });
+
+t.create('No URI specified')
+  .get('.json?query=//name&label=Package Name&style=_shields_test')
+  .expectJSON({ name: 'Package Name', value: 'no uri specified', colorB: colorsB.red });
+
+t.create('No query specified')
+  .get('.json?uri=https://services.addons.mozilla.org/en-US/firefox/api/1.5/addon/707078&label=Package Name&style=_shields_test')
+  .expectJSON({ name: 'Package Name', value: 'no query specified', colorB: colorsB.red });
+
+t.create('XML from uri')
+  .get('.json?uri=https://services.addons.mozilla.org/en-US/firefox/api/1.5/addon/707078&query=/addon/name&style=_shields_test')
+  .expectJSON({ name: 'custom badge', value: 'IndieGala Helper', colorB: colorsB.brightgreen });
+
+t.create('XML from uri (attribute)')
+  .get('.json?uri=https://services.addons.mozilla.org/en-US/firefox/api/1.5/addon/707078&query=/addon/reviews/@num')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'custom badge',
+    value: Joi.string().regex(/^\d+$/)
+  }));
+
+t.create('XML from uri | multiple results')
+  .get('.json?uri=https://services.addons.mozilla.org/en-US/firefox/api/1.5/addon/707078&query=//name')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'custom badge',
+    value: Joi.string().regex(/^.+,\s.+$/)
```

Comment:
Currently `//name` finds `IndieGala Helper, Danial Nickford, GNU General Public License, version 3.0, Firefox, Firefox for Android`. In `.+,\s.+`:
- the first part `.+` (greedy) matches `IndieGala Helper, Danial Nickford, GNU General Public License, version 3.0, Firefox`
- the last part `,\s.+` matches `, Firefox for Android`

So this regular expression does not check if all elements are separated by a comma. 

If we change query to `query=/addon/compatible_applications/application/name` we could compare result with exact value `Firefox, Firefox for Android`. What do you think about it?

Reply:
Thanks, That makes more sense.
Wasn't quite sure how to test it, so i just tried to test for the separator.

Reply:
Not sure why but the values seem to swap sides occasionally.

I ran the test 10 times and 3 of those times it failed with the values swapped around
![](https://img.shields.io/:custom%20badge-Firefox,%20Firefox%20for%20Android-blue.svg)
![](https://img.shields.io/:custom%20badge-Firefox%20for%20Android,%20Firefox-blue.svg)
what about a regex of `/^Firefox.*,\sFirefox.*$/`?

Reply:
As far as I know XPath 1.0 (`xpath` is an XPath 1.0 implementation) uses node-sets to store results, which is an unordered collection of nodes. XPath 2.0 uses node sequences, which are an ordered collections (https://www.w3.org/TR/xpath/, https://www.w3.org/TR/xpath20/). 
It would be great to use XPath 2.0 implementation, but I haven't found any good one.